File: Sorting/Ruby/merge_sort.rb
def merge_sort(arr, &prc) 
    prc ||= Proc.new { |el1, el2| el1 <=> el2 }
    return arr if (arr.length == 1 || arr.empty?)

    middle_idx = arr.length / 2
    left = merge_sort(arr.take(middle_idx), &prc)
    right = merge_sort(arr.drop(middle_idx), &prc)

    merge_sort_helper(left, right, &prc)
end 

def merge_sort_helper(arr_1, arr_2, &prc) 
    sorted_arr = [] 

    until (arr_1.empty? || arr_2.empty?)
        case (prc.call(arr_1[0], arr_2[0]))
        when -1
            sorted_arr.push(arr_1.shift)
        when 0
            sorted_arr.push(arr_1.shift)
            sorted_arr.push(arr_2.shift)
        when 1
            sorted_arr.push(arr_2.shift)
        end
    end

    arr_1.empty? ? sorted_arr += arr_2 : sorted_arr += arr_1

    return sorted_arr
end

arr = [4,6,2,3,4,7,-8,3,5,0,-6,-2,-4]

p merge_sort(arr)

File: Sorting/JavaScript/mergeSort.js
'use strict';

const minToMax = (x, y) => {
    if(x < y) {
        return -1;
    } else if(x > y) {
        return 1;
    } else if (y == x) {
        return 0;
    }
};

const maxToMin = (x, y) => {
    if(x < y) {
        return 1;
    } else if (y > x) {
        return -1;
    } else {
        return 0;
    }
};

const mergeSortHelper = (arr1, arr2, cb) => {
    let sortedArray = [];

    while (arr1.length != 0 && arr2.length != 0) {
        let n = cb(arr1[0], arr2[0]);
        if(n === -1) {
            sortedArray.push(arr1.shift());
        } else if (n === 1) {
            sortedArray.push(arr2.shift());
        } else {
            sortedArray.push(arr1.shift());
            sortedArray.push(arr2.shift());
        }
    }
    
    console.log(arr1, arr2);
    if(arr1.length === 0) {
        sortedArray = sortedArray.concat(arr2);
    } else {
        sortedArray = sortedArray.concat(arr1);
    }
    console.log(sortedArray)
    return sortedArray;
};

const mergeSort = (arr, cb = minToMax) => {
    if(arr.length == 0 || arr.length == 1) {
        return arr;
    }
    let middleIndex = parseInt(arr.length / 2)
    let leftHalf = arr.slice(0, middleIndex);
    let rightHalf = arr.slice(middleIndex, arr.length)

    let left = mergeSort(leftHalf, cb);
    let right = mergeSort(rightHalf, cb);

    return mergeSortHelper(left, right, cb)
};


let passedArr = [12, 11, 13, 5, 6, 7];
console.log(mergeSort(passedArr));